omegaconf/omegaconf.py

```diff
@@ -372,21 +372,68 @@ def _unescape_word_boundary(match: Match[str]) -> str:
 
     @staticmethod
-    def register_resolver(name: str, resolver: Resolver) -> None:
+    def register_resolver(
+        name: str,
+        resolver: Resolver,
+        args_as_strings: bool = True,
+        use_cache: Optional[bool] = True,
+    ) -> None:
+        """
+        The `args_as_strings` flag was introduced to preserve backward compatibility
+        with the older resolver system, which assumed that resolvers took the raw string
+        representation of their inputs:
+            - `True` is the old behavior (the resolver uses the string representation
+              of its inputs), and triggers a warning
+            - `False` is the new behavior (the resolver can take non-string inputs), and
+              will become the default in the future
+        """
         assert callable(resolver), "resolver must be callable"
         # noinspection PyProtectedMember
         assert (
             name not in BaseContainer._resolvers
         ), "resolved {} is already registered".format(name)
-
-        def caching(config: BaseContainer, key: str) -> Any:
-            cache = OmegaConf.get_cache(config)[name]
-            val = (
-                cache[key] if key in cache else resolver(*OmegaConf._tokenize_args(key))
-            )
-            cache[key] = val
-            return val
-
+        def caching(
+            config: BaseContainer,
+            parent: Optional[Container],
+            key: Tuple[Any, ...],
+            inputs_str: Tuple[str, ...],
+        ) -> Any:
+            # The `args_as_strings` warning is triggered when the resolver is
+            # called instead of when it is defined, so as to limit the amount of
+            # warnings (by skipping warnings when all inputs are strings).
+            if args_as_strings and any(not isinstance(k, str) for k in key):
+                non_str_arg = [k for k in key if not isinstance(k, str)][0]
+                warnings.warn(
+                    f"Resolvers that take non-string arguments should now be registered "
+                    f"with `args_as_strings=False`, and their code should be updated to "
+                    f"ensure it works as expected with non-string arguments. This "
+                    f"warning is raised because resolver '{name}' was registered with "
+                    f"the current default `args_as_strings=True` and received at least "
+                    f"one non-string argument (`{non_str_arg}`). Although we converted "
+                    f"such non-string arguments to strings to preserve backward "
+                    f"compatibility, this behavior is deprecated => please update "
+                    f"resolver '{name}' as described above. Alternatively, you may "
+                    f"ensure that all its arguments are strings, e.g., by enclosing "
+                    f"them within quotes.",
+                    category=UserWarning,
+                )
```

Comment:
1. This is a massive warning that is hard to understand.
It's best to link to an issue or doc that explains what the problem is with examples. it's difficult even for me to understand what this is telling me and what I should change, and how to ensure this and that.

2. Can you explain what this is protecting against? is this your deprecation plan for args_as_strings?

Reply:
1. A doc entry definitely sounds better, I'll do that.

2. Yes this is step 0 of the deprecation plan: tell people to update their resolvers and register them with `args_as_strings=False`

Reply:
For deprecation warnings I often create dedicated GitHub issues and link to them.
Recent example is issue 1089 in the Hydra repo, take a look.

Reply:
Just linking to related discussion for reference: https://github.com/omry/omegaconf/pull/321#discussion_r513628009

Reply:
> For deprecation warnings I often create dedicated GitHub issues and link to them.
> Recent example is issue 1089 in the Hydra repo, take a look.

Done in #426